fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnEnforcer.java

```diff
@@ -0,0 +1,109 @@
+        ColumnEnforcerVisitor visitor = new ColumnEnforcerVisitor();
+        OptExpression rootOptExpression = optExpression.getOp().accept(visitor, optExpression, enforceContext);
+        return rootOptExpression;
+    }
+
+    public List<ColumnRefOperator> getEnforcedColumns() {
+        return enforceContext.enforcedColumns;
+    }
+
+    private class EnforceContext {
+        public List<ColumnRefOperator> enforcedColumns;
+
+        public EnforceContext() {
+            this.enforcedColumns = Lists.newArrayList();
+        }
+    }
+
+    private class ColumnEnforcerVisitor extends OptExpressionVisitor<OptExpression, EnforceContext> {
+        @Override
+        public OptExpression visitLogicalTableScan(OptExpression optExpression, EnforceContext context) {
+            LogicalScanOperator scanOperator = optExpression.getOp().cast();
+            LogicalScanOperator.Builder builder = OperatorBuilderFactory.build(scanOperator);
+            builder.withOperator(scanOperator);
+            Map<ColumnRefOperator, Column> columnRefOperatorColumnMap = Maps.newHashMap(scanOperator.getColRefToColumnMetaMap());
+            for (ColumnRefOperator columnRef : columnsToEnforce) {
+                for (Map.Entry<Column, ColumnRefOperator> entry : scanOperator.getColumnMetaToColRefMap().entrySet()) {
+                    if (entry.getValue().equals(columnRef)) {
+                        columnRefOperatorColumnMap.put(columnRef, entry.getKey());
+                        context.enforcedColumns.add(columnRef);
+                    }
+                }
+            }
+            builder.setColRefToColumnMetaMap(columnRefOperatorColumnMap);
+            OptExpression newScan = new OptExpression(builder.build());
+            newScan.deriveLogicalPropertyItself();
+            return newScan;
+        }
+
+        @Override
+        public OptExpression visit(OptExpression optExpression, EnforceContext context) {
+            List<OptExpression> inputs = Lists.newArrayList();
+            EnforceContext localContext = new EnforceContext();
+            for (OptExpression child : optExpression.getInputs()) {
+                inputs.add(child.getOp().accept(this, child, localContext));
+            }
+
+            LogicalOperator.Builder builder = OperatorBuilderFactory.build(optExpression.getOp());
+            builder.withOperator(optExpression.getOp());
+
+            if (optExpression.getOp().getProjection() != null) {
+                Map<ColumnRefOperator, ScalarOperator> columnRefMap =
+                        Maps.newHashMap(optExpression.getOp().getProjection().getColumnRefMap());
+                for (ColumnRefOperator columnRef : localContext.enforcedColumns) {
+                    columnRefMap.put(columnRef, columnRef);
+                }
+                builder.setProjection(new Projection(columnRefMap));
+            }
+            OptExpression newOptExpression = OptExpression.create(builder.build(), inputs);
```

Comment:
how can we ensure all the `columnRefMap ` is valid from ScanOperator's `columnsToEnforce`? Is that possible `columnRefMap` refers non-existed column from the new ScanOperator's?

Reply:
it has been checked in line 1420-1424.